cluster-autoscaler/config/autoscaling_options.go

```diff
@@ -30,10 +30,9 @@ type GpuLimits struct {
 	Max int64
 }
 
-// AutoscalingOptions contain various options to customize how autoscaling works
-type AutoscalingOptions struct {
-	// MaxEmptyBulkDelete is a number of empty nodes that can be removed at the same time.
-	MaxEmptyBulkDelete int
+// NodeGroupAutoscalingOptions contain various options to customize how autoscaling of
+// a given NodeGroup works. Different options can be used for each NodeGroup.
+type NodeGroupAutoscalingOptions struct {
```

Comment:
I may have a dumb question but:
- Why only these 4 scale down parameters ?
`ScaleDownDelayAfterAdd`, `ScaleDownDelayAfterDelete` and so on, are not applicable for this new configuration method ?

Reply:
Those 4 values were the low-hanging fruits - they required relatively small changes in the code and they were by far the most commonly requested values to control per nodegroup.

I think ultimately it makes sense to have ScaleDownDelayAfter* as part of NodeGroupAutoscalingOptions, but it is much more tricky to do (details below) and I didn't want to block this PR on it. I'd rather take an iterative approach of gradually moving more options to NodeGroupAutoscalingOptions in separate PRs (though I don't want to make any promises on when I will be able to actually implement it for ScaleDownDelay flags; I'm also happy to just review if someone else feels like taking a shot at it).

---
Details:

ScaleDownDelayAfter* are currently implemented as [a global check](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/core/static_autoscaler.go#L496) that result in completely skipping scale-down logic] and not as a per-node check like the 4 values in the new config. Having them per-nodegroup would require completely re-implementing them and figuring out some non-obvious questions about semantics: how should per-nodegroup ScaleDownDelayAfterDelete work? It obviously blocks one nodegroup from scaling-down, but should it be triggered only when node from that nodegroup is deleted or when any node in cluster is deleted? Same goes for ScaleDownDelayAfterAdd. It seems to me that depending on use-case you may want either behavior (block on any node - if your pods can schedule on multiple nodegroups and you just want CA to pick the best one, only block on that nodegroup if you have a dedicated nodegroup per workload and each pod can generally only go to 1 nodegroup).

Reply:
OK I see, it's rather clear to me that CA should be first updated with "simple" modification first.
As for `ScaleDownDelayAfter*`, I understand the impact and the amount of source code edit behind, and the semantics questions. I may have a "dumb" solution but : what about adding a `ScaleDownAfterBehaviour`, with `global` effect and `local` or something ?
Anyway, thanks for the hard work, I really appreciate the effort to add some per node-group config so quickly, it will really boost our autoscaling configuration 🚀

Reply:
I don't think it's a dumb solution. I was considering a pretty much equivalent solution - keep the current scale-down-delay-after-* flags and add scale-down-delay-after-node-group-add/delete/etc (the name could use some work). 
This gives maximum flexibility at the cost of adding even more complexity for the user. I guess it may not be a problem if you have a hosted solution, but configuring CA is already pretty hard partly due to a huge number of flags we have. Still it may be the best solution. 

Whatever solution we decide on I'd rather do it in a separate PR though. I think this one is complex enough as is.

Reply:
> Whatever solution we decide on I'd rather do it in a separate PR though. I think this one is complex enough as is.

And I agree :)